internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,241 @@
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+	}
+}
+
+type keyType string
+
+const proxyConnectAddrKey = keyType("grpc.resolver.delegatingresolver.proxyConnectAddr")
+const userKey = keyType("grpc.resolver.delegatingresolver.user")
+
+// SetConnectAddr returns a copy of the provided resolver.Address with attributes
+// containing address to be sent in connect request to proxy.  It's data should
+// not be mutated after calling SetConnectAddr.
+func SetConnectAddr(resAddr resolver.Address, addr string) resolver.Address {
+	resAddr.Attributes = resAddr.Attributes.WithValue(proxyConnectAddrKey, addr)
+	return resAddr
+}
+
+// GetConnectAddr returns the proxy connect address in resolver.Address, or nil
+// if not present. The returned data should not be mutated.
+func GetConnectAddr(addr resolver.Address) string {
+	s, _ := addr.Attributes.Value(proxyConnectAddrKey).(string)
+	return s
+}
+
+// SetUser returns a copy of the provided address with attributes containing
+// user info. It's data should not be mutated after calling SetUser.
+func SetUser(addr resolver.Address, user *url.Userinfo) resolver.Address {
+	addr.Attributes = addr.Attributes.WithValue(userKey, user)
+	return addr
+}
+
+// GetUser returns the user info in the resolver.Address, or nil if not present.
+// The returned data should not be mutated.
+func GetUser(addr resolver.Address) *url.Userinfo {
+	user, _ := addr.Attributes.Value(userKey).(*url.Userinfo)
+	return user
+}
+
+func (r *delegatingResolver) updateState(state resolver.State) resolver.State {
+	var addresses []resolver.Address
+	for _, proxyAddr := range r.proxyAddrs {
+		for _, targetAddr := range r.targetAddrs {
+			newAddr := resolver.Address{
+				Addr: proxyAddr.Addr,
+			}
+			newAddr = SetConnectAddr(newAddr, targetAddr.Addr)
+			if r.proxyURL.User != nil {
+				newAddr = SetUser(newAddr, r.proxyURL.User)
+			}
+			addresses = append(addresses, newAddr)
+		}
+	}
+	// Set the addresses in the current state.
+	state.Addresses = addresses
+	return state
+}
+
+type wrappingClientConn struct {
+	parent       *delegatingResolver
+	resolverType string
+}
+
+// UpdateState intercepts state updates from the target and proxy resolvers.
+func (wcc *wrappingClientConn) UpdateState(state resolver.State) error {
+	wcc.parent.mu.Lock()
+	defer wcc.parent.mu.Unlock()
+	var curState resolver.State
+	if wcc.resolverType == "target" {
+		wcc.parent.targetAddrs = state.Addresses
+		curState = state
+	}
+	if wcc.resolverType == "proxy" {
+		wcc.parent.proxyAddrs = state.Addresses
+	}
+
+	if len(wcc.parent.targetAddrs) == 0 || len(wcc.parent.proxyAddrs) == 0 {
+		return nil
+	}
+	curState = wcc.parent.updateState(curState)
+	return wcc.parent.cc.UpdateState(curState)
+}
+
+// ReportError intercepts errors from the child resolvers and pass to ClientConn.
+func (wcc *wrappingClientConn) ReportError(err error) {
+	wcc.parent.cc.ReportError(err)
```

Comment:
how and where are we handling the back off?

Reply:
The delegate resolvers should be dealing with their own backoff.